stdlib/public/core/StringGraphemeBreaking.swift

```diff
@@ -177,132 +99,382 @@ extension _StringGuts {
+  internal func shouldBreak(
+    _ scalar1: Unicode.Scalar,
+    between scalar2: Unicode.Scalar,
+    _ state: inout _GraphemeBreakingState,
+    _ index: Int
+  ) -> Bool {
+    // GB3
+    if scalar1.value == 0xD, scalar2.value == 0xA {
+      return false
     }
-    return codeUnits.withUnsafeBufferPointer {
-      _measureCharacterStrideICU(of: $0, startingAt: i)
+
+    if _hasGraphemeBreakBetween(scalar1, scalar2) {
+      return true
     }
-#else
-  fatalError("No foreign strings on Linux in this version of Swift")
-#endif
-  }
 
-  @usableFromInline @inline(never)
-  @_effects(releasenone)
-  internal func _opaqueCharacterStride(endingAt i: Int) -> Int {
-    if _slowPath(isForeign) {
-      return _foreignOpaqueCharacterStride(endingAt: i)
+    let x = Unicode._GraphemeBreakProperty(from: scalar1)
+    let y = Unicode._GraphemeBreakProperty(from: scalar2)
+
+    // This variable and the defer statement help toggle the isInEmojiSequence
+    // state variable to false after every decision of 'shouldBreak'. If we
+    // happen to see a rhs .extend or .zwj, then it's a signal that we should
+    // continue treating the current grapheme cluster as an emoji sequence.
+    var enterEmojiSequence = false
+
+    defer {
+      state.isInEmojiSequence = enterEmojiSequence
     }
 
-    return self.withFastUTF8 { utf8 in
-      let (sc2, len) = _decodeScalar(utf8, endingAt: i)
-      if i &- len == utf8.startIndex {
-        // First scalar is first grapheme
-        return len
+    switch (x, y) {
+
+    // Fast path: If we know our scalars have no properties the decision is
+    //            trivial and we don't need to crawl to the default statement.
+    case (.any, .any):
+      return true
+
+    // GB4
+    case (.control, _):
+      return true
+
+    // GB5
+    case (_, .control):
+      return true
+
+    // GB6
+    case (.l, .l),
+         (.l, .v),
+         (.l, .lv),
+         (.l, .lvt):
+      return false
+
+    // GB7
+    case (.lv, .v),
+         (.v, .v),
+         (.lv, .t),
+         (.v, .t):
+      return false
+
+    // GB8
+    case (.lvt, .t),
+         (.t, .t):
+      return false
+
+    // GB9 (partial GB11)
+    case (_, .extend),
+         (_, .zwj):
+
+      // If we're currently in an emoji sequence, then extends and ZWJ help
+      // continue the grapheme cluster by combining more scalars later. If we're
+      // not currently in an emoji sequence, but our lhs scalar is a pictograph,
+      // then that's a signal that it's the start of an emoji sequence.
+      if state.isInEmojiSequence || x == .extendedPictographic {
+        enterEmojiSequence = true
+      }
+
+      return false
+
+    // GB9a
+    case (_, .spacingMark):
+      return false
+
+    // GB9b
+    case (.prepend, _):
+      return false
+
+    // GB11
+    case (.zwj, .extendedPictographic):
+      if state.isBackwards {
+        checkIfInEmojiSequence(&state, index)
+      }
```

Comment:
Emoji state is needed for both going forward and backwards. Going backwards, this emoji state lets us continuously walk further and further backwards to find the grapheme index before the emoji.

Reply:
One example of an emoji sequence that Unicode has is the following:

"A👶  🏿  ◌̈  □  👶  🏿"

==

[.any, .extendedPictographic, .extend, .extend, .zwj, .extendedPictographic, .extend]


If we start at the boundary at the last .extend, our first rule is `(.extP, .extend)` which evaluates to false, but also updates the state bit here. We keep walking until we see the A and the first emoji which is `(.any, .extP)` which is finally true.

I can change the condition to be: `if state.isBackwards, !state.isInEmojiSequence` because in this example we walk backwards for no reason because we've already evaluated that we're in an emoji sequence with the first iteration.